clang/lib/CConv/Constraints.cpp

```diff
@@ -283,17 +283,58 @@ static bool do_solve(ConstraintsGraph &CG,
   return ok;
 }
 
-auto isNonParam =
-    [](VarAtom *VA) -> bool {
-      VarAtom::VarKind VK = VA->getVarKind();
-      return VK != VarAtom::V_Param;
-    };
-
-auto isNonParamReturn =
-    [](VarAtom *VA) -> bool {
-      VarAtom::VarKind VK = VA->getVarKind();
-      return VK != VarAtom::V_Param && VK != VarAtom::V_Return;
-    };
+VarAtomPred isReturn = [](VarAtom *VA) -> bool {
+  return VA->getVarKind() == VarAtom::V_Return;
+};
+VarAtomPred isParam = [](VarAtom *VA) -> bool {
+  return VA->getVarKind() == VarAtom::V_Param;
+};
+VarAtomPred isNonParamReturn = [](VarAtom *VA) -> bool {
+  return !isReturn(VA) && !isParam(VA);
+};
+
+// Remove from S all elements that don't match the predicate P
+void filter(VarAtomPred P, std::set<VarAtom *> &S) {
+  for (auto I = S.begin(), E = S.end(); I != E; ) {
+    if (!P(*I))
+      I = S.erase(I);
+    else
+      ++I;
+  }
+}
+
+static std::set<VarAtom *> findBounded(ConstraintsGraph &CG, std::set<VarAtom *> *Concrete, bool Succs, bool UseConstAtoms) {
+  std::set<VarAtom *> Bounded;
+  std::vector<Atom *> Open;
+  if (Concrete != nullptr) {
+    Open.insert(Open.begin(), Concrete->begin(), Concrete->end());
+    Bounded.insert(Concrete->begin(), Concrete->end());
+  }
+
+  if (UseConstAtoms) {
+    auto &ConstA = CG.getAllConstAtoms();
+    Open.insert(Open.begin(), ConstA.begin(), ConstA.end());
```

Comment:
I guess I don't know what `insert` does -- what does this three-argument version do, compared to one that takes just `begin()` and `end()` as two arguments?

Reply:
Maybe it inserts at the position given by `begin()` (the first argument) ? Why would the order matter?

Reply:
Insert for `vector` requires the first argument to know where to start inserting. The only reason that's a vector is that i modeled the traversal after what is done in `do_solve`, and a vector is used there. I think I should be able to make it a `set` without breaking anything.